src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+                // node that was considered to be the top child (recursive).
+                PutAt(topChild, nodeIndex);
+                nodeIndex = topChildIndex;
+            }
+
+            PutAt(node, nodeIndex);
+        }
+
+        public partial class UnorderedItemsCollection : IReadOnlyCollection<(TElement element, TPriority priority)>, ICollection
+        {
+            private readonly PriorityQueue<TElement, TPriority> _queue;
+
+            public UnorderedItemsCollection(PriorityQueue<TElement, TPriority> queue)
+            {
+                _queue = queue;
+            }
+
+            public int Count => _queue._size;
+            object ICollection.SyncRoot => this;
+            bool ICollection.IsSynchronized => false;
+
+            public void CopyTo(Array array, int index)
+            {
+                if (array == null)
+                {
+                    throw new ArgumentNullException(nameof(array));
+                }
+
+                if (array.Rank != 1)
+                {
+                    throw new ArgumentException(SR.Arg_RankMultiDimNotSupported, nameof(array));
+                }
+
+                if (array.GetLowerBound(0) != 0)
+                {
+                    throw new ArgumentException(SR.Arg_NonZeroLowerBound, nameof(array));
+                }
+
+                if (index < 0 || index > array.Length)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(index), index, SR.ArgumentOutOfRange_Index);
+                }
+
+                if (array.Length - index < _queue._size)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidOffLen);
+                }
+
+                try
+                {
+                    Array.Copy(_queue._nodes, 0, array, index, _queue._size);
+                }
+                catch (ArrayTypeMismatchException)
+                {
+                    throw new ArgumentException(SR.Argument_InvalidArrayType, nameof(array));
+                }
+            }
+
+            public struct Enumerator : IEnumerator<(TElement element, TPriority priority)>, IEnumerator
+            {
+                private readonly PriorityQueue<TElement, TPriority> _queue;
+                private readonly int _version;
+
+                private int _index;
+                private (TElement element, TPriority priority)? _currentElement;
+
+                private const int FirstCallToEnumerator = -2;
+                private const int EndOfEnumeration = -1;
+
+                internal Enumerator(PriorityQueue<TElement, TPriority> queue)
+                {
+                    _queue = queue;
+                    _version = queue._version;
+                    _index = FirstCallToEnumerator;
+                    _currentElement = default;
+                }
+
+                public void Dispose()
+                {
+                    _index = EndOfEnumeration;
```

Comment:
This needs to release _currentElement.

Reply:
OK, strictly it's undefined and we're not consistent -- Queue<T> does but others don't. It seems like releasing is a good pattern.

Reply:
In fact, the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerator-1?view=net-5.0) suggest that unless there's some resource to dispose, it should be empty. That seems odd as it doesn't protect against the type being used post dispose, which makes it hard to implement dispose later.

Reply:
Most enumerators do nothing in Dispose, including most implemented by the compiler for iterators (Dispose will nop unless there's something it actually needs to do, like handle finallys).

Reply:
OK, then this should be empty then.